src/System.Diagnostics.Tracing.Telemetry/src/System/Diagnostics/Tracing/TelemetryListener.cs

```diff
@@ -0,0 +1,264 @@
+        /// because they are in a global list (for discoverability).  You must dispose them explicitly. 
+        /// Note that we do not do the Dispose(bool) pattern because we frankly don't want to support
+        /// subclasses that have non-managed state.   
+        /// </summary>
+        virtual public void Dispose()
+        {
+            // Remove myself from the list of all listeners.  
+            lock (DefaultListener)
+            {
+                if (_disposed)
+                    return;
+                _disposed = true;
+                if (s_allListeners == this)
+                    s_allListeners = s_allListeners._next;
+                else
+                {
+                    var cur = s_allListeners;
+                    while (cur != null)
+                    {
+                        if (cur._next == this)
+                        {
+                            cur._next = _next;
+                            break;
+                        }
+                        cur = cur._next;
+                    }
+                }
+                _next = null;
+            }
+
+            // Indicate completion to all subscribers.  
+            Subscription subscriber = null;
+            Interlocked.Exchange(ref subscriber, _subscriptions);
+            while (subscriber != null)
+            {
+                subscriber.Observer.OnCompleted();
+                subscriber = subscriber.Next;
+            }
+            // The code above also nulled out all subscriptions. 
+        }
+
+        /// <summary>
+        /// When a TelemetryListener is created it is given a name.   Return this.  
+        /// </summary>
+        public string Name { get; private set; }
+
+        /// <summary>
+        /// Return the name for the ToString() to aid in debugging.  
+        /// </summary>
+        /// <returns></returns>
+        public override string ToString()
+        {
+            return Name;
+        }
+
+        #region private
+
+        // NotificationSource implementation
+        /// <summary>
+        /// Override 
+        /// </summary>
+        public override bool IsEnabled(string telemetryName)
+        {
+            for (Subscription curSubscription = _subscriptions; curSubscription != null; curSubscription = curSubscription.Next)
+            {
+                if (curSubscription.IsEnabled == null || curSubscription.IsEnabled(telemetryName))
+                    return true;
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Override 
+        /// </summary>
+        public override void WriteTelemetry(string telemetryName, object parameters)
+        {
+            for (Subscription curSubscription = _subscriptions; curSubscription != null; curSubscription = curSubscription.Next)
+                curSubscription.Observer.OnNext(new KeyValuePair<string, object>(telemetryName, parameters));
```

Comment:
Here is the code for the body of the loop in a X64 release build 

                curSubscription.Observer.OnNext(new KeyValuePair<string, object>(telemetryName, arguments));
00007FFBC0D4C342  mov         rcx,qword ptr [rbx+8]  
00007FFBC0D4C346  lea         rdx,[rsp+20h]  
00007FFBC0D4C34B  mov         qword ptr [rdx],rsi  
00007FFBC0D4C34E  mov         qword ptr [rdx+8],rdi  
00007FFBC0D4C352  lea         rdx,[rsp+20h]  
00007FFBC0D4C357  mov         r11,7FFBC0A510E8h  
00007FFBC0D4C361  cmp         dword ptr [rcx],ecx  
00007FFBC0D4C363  call        qword ptr [r11]  

The call at the bottom is the OnNext() call and the movs at 34B and 34E are the setup of the keyValuePair.    There is a redundant LEA instruction but otherwise it looks pretty good.